clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java

```diff
@@ -68,13 +86,32 @@ public String source() {
 
     @Override
     public boolean complete() {
-        return !size.hasRemaining() && !buffer.hasRemaining();
+        return !size.hasRemaining() && buffer != null && !buffer.hasRemaining();
     }
 
     public long readFrom(ScatteringByteChannel channel) throws IOException {
         return readFromReadableChannel(channel);
     }
 
+    @Override
+    public boolean requiredMemoryAmountKnown() {
+        return requestedBufferSize != -1;
+    }
+
+    @Override
+    public boolean memoryAllocated() {
+        return buffer != null;
+    }
+
+
+    @Override
+    public void close() throws IOException {
+        if (buffer != null && buffer != EMPTY_BUFFER) {
+            memoryPool.release(buffer);
+            buffer = null;
```

Comment:
Since on the server side, we release memory through RequestChannel.dispose(). However, memory is also released here through KafkaChannel.close(). Will this cause the same memory to be released more than once in certain cases?